staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go

```diff
@@ -62,112 +65,134 @@ type result struct {
 type wrappedLifecycleSignal struct {
 	lifecycleSignal
-	callback func(bool, string, lifecycleSignal)
+	before func(lifecycleSignal)
+	after  func(lifecycleSignal)
 }
 
 func (w *wrappedLifecycleSignal) Signal() {
-	var name string
-	if ncw, ok := w.lifecycleSignal.(*namedChannelWrapper); ok {
-		name = ncw.name
-	}
-
-	// the callback is invoked before and after the termination event is signaled
-	if w.callback != nil {
-		w.callback(true, name, w.lifecycleSignal)
+	if w.before != nil {
+		w.before(w.lifecycleSignal)
 	}
 	w.lifecycleSignal.Signal()
-	if w.callback != nil {
-		w.callback(false, name, w.lifecycleSignal)
+	if w.after != nil {
+		w.after(w.lifecycleSignal)
 	}
 }
 
-func wrapLifecycleSignals(t *testing.T, ts *lifecycleSignals, callback func(bool, string, lifecycleSignal)) {
-	newWrappedLifecycleSignal := func(delegated lifecycleSignal) lifecycleSignal {
-		return &wrappedLifecycleSignal{
-			lifecycleSignal: delegated,
-			callback:        callback,
-		}
-	}
+func wrapLifecycleSignalsWithRecorer(t *testing.T, signals *lifecycleSignals, after func(lifecycleSignal)) {
+	signals.AfterShutdownDelayDuration = wrapLifecycleSignal(t, signals.AfterShutdownDelayDuration, nil, after)
+	signals.PreShutdownHooksCompleted = wrapLifecycleSignal(t, signals.PreShutdownHooksCompleted, nil, after)
+	signals.NotAcceptingNewRequest = wrapLifecycleSignal(t, signals.NotAcceptingNewRequest, nil, after)
+	signals.HTTPServerStoppedListening = wrapLifecycleSignal(t, signals.HTTPServerStoppedListening, nil, after)
+	signals.InFlightRequestsDrained = wrapLifecycleSignal(t, signals.InFlightRequestsDrained, nil, after)
+	signals.ShutdownInitiated = wrapLifecycleSignal(t, signals.ShutdownInitiated, nil, after)
+}
 
-	ts.AfterShutdownDelayDuration = newWrappedLifecycleSignal(ts.AfterShutdownDelayDuration)
-	ts.HTTPServerStoppedListening = newWrappedLifecycleSignal(ts.HTTPServerStoppedListening)
-	ts.InFlightRequestsDrained = newWrappedLifecycleSignal(ts.InFlightRequestsDrained)
-	ts.ShutdownInitiated = newWrappedLifecycleSignal(ts.ShutdownInitiated)
+func wrapLifecycleSignal(t *testing.T, delegated lifecycleSignal, before, after func(_ lifecycleSignal)) lifecycleSignal {
+	return &wrappedLifecycleSignal{
+		lifecycleSignal: delegated,
+		before:          before,
+		after:           after,
+	}
 }
 
-type step struct {
-	waitCh, doneCh chan struct{}
-	fn             func()
+type testStep struct {
+	doneCh chan struct{}
 }
 
-func (s step) done() <-chan struct{} {
-	close(s.waitCh)
-	return s.doneCh
+func (ts testStep) done() <-chan struct{} {
+	return ts.doneCh
 }
-func (s step) execute() {
-	defer close(s.doneCh)
-	<-s.waitCh
-	s.fn()
+func (ts testStep) execute(fn func()) {
+	defer close(ts.doneCh)
+	fn()
 }
-func newStep(fn func()) *step {
-	return &step{
-		fn:     fn,
-		waitCh: make(chan struct{}),
+func newTestStep() *testStep {
+	return &testStep{
 		doneCh: make(chan struct{}),
 	}
 }
 
+//  This test exercises the graceful termination scenario
```

Comment:
i think github UI does some magic with whitespaces, but on my editor the diagram looks fine.

Reply:
It seems that you are using "tabs`" instead of whitespaces.

Please switch to using whitespace and it should get fixed...

Reply:
Here it is :)